sotodlib/data/load.py

```diff
@@ -0,0 +1,317 @@
+                comp_info = (gain, offset)
+            else:
+                comp_info = None
+            target = fo[item.name]
+            unpack_frame_object(target, item, streams[item.name], comp_info)
+            if item.timestamp_field is not None:
+                t0, t1, ns = target.start, target.stop, target.n_samples
+                t0, t1 = t0.time / core.G3Units.sec, t1.time / core.G3Units.sec
+                streams[item.timestamp_field].append(np.linspace(t0, t1, ns))
+            continue
+        # This is a simple field.
+        if compression_info is not None:
+            gain, offset = compression_info
+            m, b = gain.get(item, 1.), offset.get(item, 0.)
+            v = np.array(fo[item], dtype='float32') / m + b
+        else:
+            v = np.array(fo[item])
+        streams[item].append(v)
+
+def unpack_frames(filename, field_request, streams):
+    """Read frames from the specified file and expand the data by stream.
+    Only the requested fields, specified through *_fields arguments,
+    are expanded.
+
+    Arguments:
+      filename (str): Full path to the file to load.
+      field_request: Instructions for what fields to load.
+      streams: Structure to which to append the
+        streams from this file (perhaps obtained from running
+        unpack_frames on a preceding file).
+
+    Returns:
+      streams (structure containing lists of numpy arrays).
+
+    """
+    if streams is None:
+        streams = field_request.empty()
+    
+    reader = so3g.G3IndexedReader(filename)
+    while True:
+        frames = reader.Process(None)
+        if len(frames) == 0:
+            break
+        frame = frames[0]
+        if frame.type == core.G3FrameType.Scan:
+            unpack_frame_object(frame, field_request, streams)
+    return streams
+
+def load_observation(db, obs_name, dets=None, prefix=None,
+                     mpicomm=None, toast=False):
+    """Load the data for some observation.  You can restrict to only some
+    detectors. Coming soon: also restrict by time range / sample
+    index.
+
+    This specifically targets the pipe-s0001 sim format.
+
+    Arguments:
+
+      db (ObsFileDB): The database describing this observation file
+        set.
+      obs_name (str): The identifier of the observation.
+      dets (list of str): The detector names of interest.  If None,
+        loads all dets present in this observation.
+      prefix (str): The root address of the data files.  If not
+        specified, the prefix is taken from the ObsFileDB.
+
+    Returns:
+      (signal_streams, other_streams).
+
+    """
+    if prefix is None:
+        prefix = db.prefix
+        if prefix is None:
+            prefix = './'
+
+    # Regardless of what dets have been asked for (maybe none), get
+    # the list of detsets implicated in this observation.
+    c = db.conn.execute('select distinct DS.name, DS.det from detsets DS '
```

Comment:
This DB query is what could be used when loading data into toast, rather than the current traversal of the filesystem to find observations:  https://github.com/simonsobs/sotodlib/blob/toast_refactor/sotodlib/data/toast_load.py#L574